### ISSUE TITLE
Improve the file upload widget 

### ISSUE THREAD

**Issue by frjo** [OpenTechFund/hypha#1440]
- [x] An drag and drop area
- [x] Show a list of the uploaded files
- [x] Allow user to add files to the list
- [x] Allow user to remove files from the list

Something like this:

![Skärmavbild 2019-08-21 kl  10 12 14](https://user-images.githubusercontent.com/191392/63414475-37fccc80-c3fc-11e9-8840-7868705304ec.png)

(from commercial https://innostudio.de/fileuploader/)

My first option would be to look for an open source javascript widget that works with the file upload stream field we have now.

**Comment by frjo** [comment]
@Parbhat The current file upload widget is not nice for users. I would like something that looks and works the way users are used to from other modern sites.

**Comment by Parbhat** [comment]
@frjo Looked into different open source file uploaders and seems that we have to change the way the form submits. Most Drag and drop implementations require the form submission via AJAX instead of the normal form submission (current). Because we cannot manipulate or populate native HTML file input using Javascript. So, to include the modern file upload widgets we need some changes in the form submission process. The new process should also take into account some current form side effects like creating a user account if not present.

Quotes from different sources:

>There is no cross-browser way to upload dragged & dropped files without Ajax. Some browsers (IE and Firefox) do not allow setting the value of a file input, which then could be submitted to server in a usual way. - [css-tricks](https://css-tricks.com/drag-and-drop-file-uploading/#article-header-id-4)

>It's not possible to assign files to an <input type="file"> element, and Uppy is built around that assumption. Many of our features like restrictions and remote provider uploads cannot be implemented while supporting plain `<form>` uploads. - [uppy](https://github.com/transloadit/uppy/issues/872)

>The primary reason for using AJAX mode for drag and drop is that you cannot manipulate or populate native HTML file input using external script or javascript (it is meant to be readonly for javascript assignment and no files cannot be assigned). - [bootstrap-fileinput](https://github.com/kartik-v/bootstrap-fileinput/issues/818#issuecomment-268844452)

**Comment by frjo** [comment]
We can test https://github.com/mbraak/django-file-form

**Comment by fourthletter** [comment]
Please deploy if the test works. Thanks!

**Comment by saoun** [comment]
@todd-dembrey is this something that you can take a look at and help out with?

**Comment by rudietuesdays** [comment]
@saoun @todd-dembrey @fourthletter @frjo wondering what the status is on this since I don't see any commits associated? Wondering if it's in progress somewhere already or still needed to start from the top.

**Comment by frjo** [comment]
@rudietuesdays We have asked Torchbox (i.e. @todd-dembrey) to take this on. But if you feel up to it your are welcome to take it on.

Users really want this issue resolved, the current widget is non intuitive and cause duplicated from time to time.

**Comment by abigail-hampson** [comment]
Hello,

Could you please let us know the issues/bug that there is with the current set up and any examples or screenshots of this? 

Thanks, 
Abigail

**Comment by frjo** [comment]
1. User can not use drag and drop, not for a single file and not for multiple files. Single files will load in the browser, replacing the form.
2. When user have added file there is no way to remove it. Except by uploading another file, new selection will replace all current ones.
3. It's not possible to add one file and then add another. You need to pick all the needed files in one go.

In short, it doesn't look or work like most other file upload fields on the web.

Go to https://innostudio.de/fileuploader/#examples and pick "Ajax" version. That is more or less what we want. Not that exact look, but that functionality.

**Comment by balazs-endresz** [comment]
Hi all, I just started to look into this for Torchbox and I'd suggest implementing only a non-ajax version.

We've tried the `Default` example on https://innostudio.de/fileuploader/#examples and all features (multi upload, dnd, remove) work in Chrome, Firefox, Safari but not in IE/Edge. As far as I can tell this wouldn't require many changes to the server side form processing code (maybe none), unlike the ajax uploader. It would save likely several days of work, while keeping the codebase much simpler / easier to maintain.

On IE/Edge we'd still have the current file input as a fallback (and different instructions for users) but everyone else could use the enhanced version.

The `Default` example on https://innostudio.de/fileuploader/#examples can be easily improved by increasing the drop area with css, and/or adding a handler to prevent navigation if a file is dropped outside of it.

Also, currently files that were previously uploaded are listed separately from the upload field: https://i.imgur.com/mbT24ZS.png The easiest thing to do would be to leave that as is and use the new upload widget only when uploading new files. But this is something we could improve on further to have a more unified look. At least make it work like this: https://dev-apply.opentech.fund/apply/projects/payment-requests/2/edit/

What do you think?

**Comment by frjo** [comment]
Hi @balazs-endresz,

1. IE 11 support is not needed anymore and we will be removing any IE 11 hacks/code, see https://github.com/OpenTechFund/hypha/issues/2037. Only support for Edge with Chromium engine is needed and that should not need anything extra I believe.
2. If possible I would like to remove current file listings, they are a bit of a hack. A new clean and uniform solution to uploads fields that are the same throughout the app would be much appreciated by all parties. But as an interim solution we can keep them if really needed.
3. Just to make it clear, the mention of the fileuploader solution above is just an example. It was simply the first page we found that looked something like what we wanted.

**Comment by balazs-endresz** [comment]
1. So, **would you mind keeping the current simpler native file upload UI on Edge**? And add better ones only for Chrome, etc?
2. Some file upload widgets probably support handling already uploaded files too (I think it's the `Preloaded files` section for the innostudio one), but that would mean changing a few things on the server side to be able to list and delete them in a way that's expected by the widget. So, it's likely easier to just make the styling more consistent instead.
3.
 * Wagtail uses (an older version of) jQuery-File-Upload: https://blueimp.github.io/jQuery-File-Upload/ 
 * https://www.dropzonejs.com/ also seems to be popular and doesn't need jquery
 * https://innostudio.de/fileuploader/ for a €10 lifetime license doesn't sound bad either.  It's not nearly as popular but they have been active recently: https://github.com/innostudio/fileuploader/issues?q=is%3Aissue+is%3Aclosed It looks a bit more polished out of the box and we already tried it in various browsers, so I'm slightly leaning towards this. But if there's an issue with this it should be easy to switch to any other library later (assuming we do it without ajax).

**Comment by frjo** [comment]
1. If it works on Google Chrome does it not work on any Chromium based browser, like latest Edge? Old versions of IE and Edge you can safely ignore.
2. Ok, good argument.
3. 10 USD is nothing but it would be something each user of Hypha would need to pay and that I really would like to avoid. All code in Hypha should be open source.

**Comment by balazs-endresz** [comment]
I was looking at legacy Edge but it does work on the Chromium based one.

I'll try adding Dropzone or jQuery-File-Upload then.

**Comment by balazs-endresz** [comment]
I've tried both dropzone.js and jquery-file-upload on `/apply/projects/reports/4/edit/` but turns out neither of those can do non-ajax uploads. Unfortunately I couldn't even find any other ones that do. I guess because until recently ajax-upload was the only viable solution because of browser support.

It's worth mentioning that the innostudio one creates multiple file input fields, all with `name="file[]"`. Perhaps we could try to replicate that from scratch but it's hard to tell how complex it would be. However, ajax uploads are not necessarily easier either, especially given that file uploads are done in various different context in this project (e.g. generated from streamfield).

**Comment by BoPeng** [comment]
Just want to comment that django-file-form made a new release recently with support for drag & drop box, placeholder for existing file, directory upload etc.

**Comment by frjo** [comment]
@balazs-endresz What is the way forward for this? Getting the file upload fields in Hypha to work like they do in most other sites is a very hight priority. Users constently have problems with the current implementation.

**Comment by balazs-endresz** [comment]
I've looked at django-file-form briefly at the beginning but I saw that it used the local file system directly, instead of the django storage backend, so that might be problem with heroku and S3. It also requires customising form classes, which may be a problem with forms generated by streamfield. Unfortunately there's no demo either, so we'd need more time investigating that as an option. At least it does claim to support multiple file upload without ajax, however I don't see an option to configure it, so that might be referring to the native fallback only.

Basically we can have either ajax or non-ajax uploads and both of those are difficult to estimate and likely take several days if not more than a week to implement. (Not a major issue but I also had trouble getting the frontend tooling working properly, which adds to it as well.) 

But if you don't mind us going ahead with this then first I'd try to see if django-file-form actually supports non-ajax uploads. If that is not suitable then spend some time replicating what the innostudio uploader does (i.e. multiple hidden file input fields, which require no server side changes). This might be overly ambitious but perhaps given the browser support it's not that difficult to do now. Failing that we could do ajax uploads, preferably with django-file-form or something similar but if those don't work out of the box then we might have to do the server side implementation from scratch.

**Comment by frjo** [comment]
@balazs-endresz  Please go ahead! Hypha need a modern field upload widget.

The main concerns we have is user experience and that the solution is maintainable and easy for other developers to grasp.

If wagtail uses jQuery-File-Upload that would be ok for Hypha as well, we have jQuery loaded already.

**Comment by frjo** [comment]
Does not Torchbox have a solution that you use on other sites?

**Comment by abigail-hampson** [comment]
Hi @saoun , 

Can I just check you're happy with this from a budget perspective too as this will be a considerable chunk of work? 

Thanks, 
Abigail

**Comment by BoPeng** [comment]
We investigated file upload issue for a while and though AJAX-upload is the way to go. Unfortunately only `django-file-form` provides customized fields to bridge AJAX frontend and `django` form (we do not use streamfield).

`django-file-form` does multi-file upload nicely but did not have a drop widget. We worked with its author for a while and added a few features for our application (mostly file drop and editing existing file list). The biggest problem right now is that it uses local (webserver) for download so the webserver has to process the uploaded files and upload them to S3, which can be a problem for a busy webserver. 

We therefore investigated uppy, which supports direct upload to S3, but does not work well with django form. The author of django-file-form said that direct-to-S3 is out of the scope. This is why I noticed your discussions. 

If everything fails, maybe we can fork `django-file-form` and replace its frontend widget with `uppy`. Perhaps @balazs-endresz is interested in participating?

**Comment by frjo** [comment]
@balazs-endresz How is the work going?

@BoPeng Thanks for adding your input to our problem! If practical I would like to use django-file-form instead of us inventing the wheel again and we have no problem throwing money at the problem.

**Comment by abigail-hampson** [comment]
Hello, Balazs isn't on this until late next week. So we'll be picking this up then. Is there anything we need to sort or agree on before that @balazs-endresz ?

**Comment by balazs-endresz** [comment]
> The biggest problem right now is that it uses local (webserver) for download so the webserver has to process the uploaded files and upload them to S3, which can be a problem for a busy webserver.

Given that we don't do direct upload to S3 at the moment I assume that isn't a priority here. 

> If practical I would like to use django-file-form instead of us inventing the wheel again and we have no problem throwing money at the problem.

Django-file-form is likely not a drop in solution (for example because it doesn't seem to use the django storage backend) but worth a try. And Ben mentioned the `MultiFileFieldBlock` which does file uploads in a streamfield based form. **Would you need the same improved upload widget for that too?** I might be wrong but I suspect that could make it a bit more complex to add django-file-form (or any ajax upload solution).

Anyway, I think first I'd spend only a few hours to see if we could replicate what the innostudio uploader does. Essentially it's just adding a new file input element each time a file is added, which can handle drag and drop natively.

Btw, we've just added dropzone.js to another project, which was even simpler in some ways (e.g. had a separate page dedicated for file uploads only / not using S3) and that took several days too.

**Comment by frjo** [comment]
@balazs-endresz It is for the streamfields we need the improvements. These are the ones used on all the submission and project forms. (The standard image and file uploads in Wagtail admin works just fine, no need to do anything with them.)

(In a separate note I which FileField and MultiFileField was merged in to one field type.)

**Comment by balazs-endresz** [comment]
Hi all

I've ended up trying django-file-form first. It wasn't painless (e.g. turns out doesn't work at all with the DummyCache) but got it working locally. 

I saw that they put files in the database right after it's uploaded: https://github.com/mbraak/django-file-form/blob/7fdc51c44a5bdf4cca37f5937156aaf272a012b7/django_file_form/views/tus.py#L187
But they still keep files in the local filesystem afterwards too, and if those are removed then we get 500 errors, which I don't really understand. Maybe it's just a performance optimisation. So, this might work most of the time but will cause error when dynos are restarted.

**Comment by balazs-endresz** [comment]
Ah, sorry, I posted it early by accident. I haven't fully understood the code yet but basically this does seem like a major issue to me with django-file-form.

**Comment by balazs-endresz** [comment]
> But they still keep files in the local filesystem afterwards too, and if those are removed then we get 500 errors, which I don't really understand. Maybe it's just a performance optimisation. 

So, `UploadedFile` does not store the file contents in the database: https://github.com/mbraak/django-file-form/blob/7fdc51c44a5bdf4cca37f5937156aaf272a012b7/django_file_form/models.py#L45 

But the we still have the same issue: 500 errors if a dyno is restarted between a file upload and a form submission. That's not something you'd like to put up with I assume?

**Comment by BoPeng** [comment]
@balazs-endresz I do not fully understand the code either but I believe the database is used for the tus-interface that allows resumable upload. After the files are uploaded, it provides a `File` interface with content through form `FILE`. The files in the database will cleaned after 1 day and through the [clean unused file command](https://github.com/mbraak/django-file-form/blob/master/django_file_form/management/commands/delete_unused_files.py)

**Comment by balazs-endresz** [comment]
I don't think the database is used for the resumable uploads: it creates an `UploadedFile` only after a file is fully uploaded.

Anyway, the main issue is that dyno restarts between file upload and form submission will cause an error. But I'll see if there's a way around that. 

Sorry again, I hit Ctrl+Enter a bit too early by accident, I was still thinking this through :)

Also, coming back to the non-ajax uploads: **if the form validation fails that will mean users have to add all the files again. Would that be a problem?** In contrast, django-file-form handles that nicely.

**Comment by frjo** [comment]
Dyno reload between file upload and form submission I think would be rare indeed.

Disappearing files after failed form validation is something that I think would confuse the users and create frustration.

**Comment by balazs-endresz** [comment]
> Dyno reload between file upload and form submission I think would be rare indeed.

Dynos are restarted _at least_ once every day and on every deployment too: https://devcenter.heroku.com/articles/dynos#restarting

> Disappearing files after failed form validation is something that I think would confuse the users and create frustration.

Form validation works great with django-file-form but the problem is only when there's a dyno restart between the file upload and a _successful_ form submission.

But I can add django-file-form on a branch and you can give it a try.

Alternatives to using django-file-form as is (i.e. with the dyno restart issue) would be:

* Forking it and uploading even _temporary_ files to S3. This will cause some extra traffic but django-file-form integrates with django nicely, so I think this would be reasonable. This wouldn't be *direct* upload to S3 from the browser but through heroku. And now I see even streamfield based forms have a custom form class defined already, so that shouldn't cause any extra issues (I was wary of this before).
* Create the temporary files on S3 but adding the code for this to our forms instead. This would essentially replace some (or perhaps a lot) of the functionality of django-file-form, and increase complexity. This wouldn't be my first choice.
* Non-ajax upload: I didn't notice this before but the main UX issue with this is that **after a failed form validation newly added files would have be re-uploaded. Do you think that would be acceptable?** We could probably make this less likely with more client side form validation.

**Comment by balazs-endresz** [comment]
If I change only the storage backend for django-file-form: https://github.com/mbraak/django-file-form/blob/7fdc51c44a5bdf4cca37f5937156aaf272a012b7/django_file_form/models.py#L45 that seems to be enough to avoid the dyno restart issue. So far, I've tried it with a simple custom storage backend but I'll set up an S3 bucket for my local build and see if that still works. (This is fairly easy to do with https://github.com/torchbox/buckup )

This way django-file-form could keep using the local filesystem while it's uploading files but as soon as a file is fully uploaded it creates an `UploadedFile` instance, which will now have a FileField with the S3 storage. Whenever that FileField is accessed that's already done via the django storage backend it seems (not the raw filesystem api), so this shouldn't require any further code changes to django-file-form.

The temporary files will remain on the local filesystem too but the daily dyno restart will take care of those. The `delete_unused_files` won't work out of the box but one can just delete `UploadedFile` instance simply from the django admin instead. 

I'll add another PR after I tried it with S3. And if this works out then hopefully this change could be merged back to django-file-form too.

**Comment by frjo** [comment]
Thanks for working on this! Sounds like you are making great progress. Will test your wip pr this evening. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BoPeng** [comment]
`django-file-form` had an option for customized storage, but it was [removed](https://github.com/mbraak/django-file-form/pull/280) because [it did not work properly](https://github.com/mbraak/django-file-form/issues/277). Please consider submitting a PR if @balazs-endresz has made it work properly.

**Comment by balazs-endresz** [comment]
Oh, that's interesting. I've just added a new MR here using a django-file-form fork that seems to work for me: https://github.com/OpenTechFund/hypha/pull/2086

If you have no issues with that it would be definitely good to get it merged back to django-file-form.

**Comment by balazs-endresz** [comment]
I've added it to the stream forms on the same branch too but couldn't try it locally because it's rendered only for the preview not when viewing the actual page, even though I've published it.

@frjo Could you check why I can't log in on dev? I've sent details via email. And can I really merge this to `develop` now? That won't be accidentally merged to `master` before it's all done?

**Comment by balazs-endresz** [comment]
There were some weird issues with static files and heroku but the dev server is back up now.

But for example on https://dev-apply.opentech.fund/rapid-response-fund/ I'm getting a 403 error when it starts to upload a file with a PATCH request. This doesn't show up in the heroku logs and the response mentions captchas and cloudflare so we'll need to change some settings on CF. First we could try to make the `/upload/upload/*` path exempt just to see if uploads work otherwise. But I'm not sure if that's a proper solution.

**Comment by frjo** [comment]
@balazs-endresz I turned the web application firewall off in Cloudflare for the dev site now.

Tested to upload a number of files to a new test submission on the dev site and it worked perfectly. I can also confirm that the files ended up in the S3 bucket as they should. Really nice work!

The part that remains is handling editing submissions, now no files show.

**Comment by BoPeng** [comment]
> The part that remains is handling editing submissions, now no files show.

This is handled by creating placeholder files during form initialization. The "Adding placeholder files" section of the [`jango-file-form` README](https://github.com/mbraak/django-file-form) describes it.

**Comment by balazs-endresz** [comment]
> > The part that remains is handling editing submissions, now no files show.
> 
> This is handled by creating placeholder files during form initialization. The "Adding placeholder files" section of the [`jango-file-form` README](https://github.com/mbraak/django-file-form) describes it.

I haven't looked into that before but I guess we might have to deal with `StreamBaseForm.add_placeholder_inputs` then. I had to disable that method for streamforms only. I also mentioned this in the list above.  Maybe the problem was caused by the custom metaclass because that broke `form.hidden_inputs` as well.

**Comment by BoPeng** [comment]
Yeap, `django-file-form` use a [hidden field](https://github.com/mbraak/django-file-form/blob/master/django_file_form/widgets.py#L51) to keep track of placeholders.

**Comment by balazs-endresz** [comment]
I think I fixed most of the backend issues but haven't been able to test the stream forms locally and on dev I saved this as draft with a dummy pdf file: https://dev-apply.opentech.fund/rapid-response-fund/ but I'm not sure how to check what happened to it.

**Comment by balazs-endresz** [comment]
**QUESTION:**

* [ ] Can I fork django-file-form under https://github.com/OpenTechFund and open a PR from there? I guess that would make it easier for others to make changes to it. If it doesn't make it to the official release in time you should install it from your own repo anyway, and not from my personal one.

**Comment by frjo** [comment]
@balazs-endresz I have created an fork here https://github.com/OpenTechFund/django-file-form

Please add your patch to that branch and then we can create an PR from there.

**Comment by frjo** [comment]
This is what got uploaded to S3, looks like it works.

![dev-s3-uploads](https://user-images.githubusercontent.com/191392/87778306-4aa99c80-c82b-11ea-904f-f179efcf77dc.png)

**Comment by frjo** [comment]
@Parbhat Can your team take over this and work on the remaining issues?

This is one of the few things that are as important as the API work.

Things that remains that I know of are:

- [ ] Add the feature to submission edit forms, now only works when you add a new submission.
- [ ] Client side validation, users should be told directly if a non valid file is uploaded e.g.
- [ ] Deleting files that have been uploaded seems not to work. Create a new submission, add a JPEG file and try to submit. Form validation will complain that JPEGs are not allowed, clicking the delete link will remove the file in the UI but when trying to submit again it return.

As a first step we can concentrate on the submission (application) form upload stream fields. The same widget should be used everywhere but we can do that later if it's time consuming. The application forms is most important by far.

**Comment by Parbhat** [comment]
@frjo Yeah, assigned @sks444 so he can start work on the remaining issues.

**Comment by sks444** [comment]
@frjo, along with the issues that you mentioned.

- [ ] `django-file-form` is not yet implemented to be used for single file uploads stream block, I believe that needs to be done as well.

**Comment by frjo** [comment]
@sks444 Yes, that needs to be implemented. Thanks for noticing that!

**Comment by sks444** [comment]
I have made some progress on this, most of the functionalities are working fine now, except that there are few bugs that needs to be fixed when dealing with edit forms, one of which also needs to be fixed at django-file-form https://github.com/mbraak/django-file-form/issues/318(I will send in a pr for that). And then, In case of multi file upload,  when user edits them I would need to get the original files for the placholders and update the db accordingly.

Other then that one client side validation of the invalid files would be remaining, would be good to have that functionality at https://github.com/mbraak/django-file-form/issues/317 but in the meantime we could try to fix the current solution that we have at https://github.com/OpenTechFund/hypha/blob/d80077223db97347f3a2c7a8e2510f56efc7c6c7/hypha/static_src/src/javascript/apply/file-uploads.js#L20.

**Comment by frjo** [comment]
Excellent progress on this @sks444! Will test and see if I can find any other edge cases.

**Comment by frjo** [comment]
@sks444 Deleting files that fails validation (e.g. a JPEG) now works. Viewing files on a submission works. Adding files when editing an submission works. Things remaining that I found is:

- [ ] Works for not logged in users on submission (application) forms.
- [ ] Remove files when editing a submission. There is a deleting link and it looks like they are gone but after save they are still there.

**Comment by frjo** [comment]
@sks444 @Parbhat  Noticed something that looks like a stray label tag in:

https://github.com/OpenTechFund/hypha/blob/6d446e4fb577dbb4a5069121cf353c91d255b8c4/hypha/apply/templates/forms/includes/field.html#L43

Does it has a purpose or can we just remove it?

**Comment by sks444** [comment]
Yeah @frjo, I have fixed those issues in the last commit which I just pushed.

**Comment by Parbhat** [comment]
@frjo It looked like a leftover initially but required for `checkbox_input` widget type after testing on live and checking differences in https://github.com/OpenTechFund/hypha/commit/4a4986908e3f7eb65a83107e9e535ae781c19dbc. We can have a condition check for checkbox input widget and include it only for it. As other input types do not need it.

**Comment by fourthletter** [comment]
I attempted to upload files to this application https://test-apply.opentech.fund/apply/submissions/260/ and it failed. 

<img width="822" alt="Screen Shot 2020-08-04 at 9 44 07 AM" src="https://user-images.githubusercontent.com/20019656/89301109-380ad200-d637-11ea-9b6a-ea52c41fdce4.png">

**Comment by frjo** [comment]
@fourthletter My bad. Forgot to adjust the Cloudflare firewall. Fixed now so please retest.

**Comment by fourthletter** [comment]
As an applicant I was able to upload documents, I could upload a single document, upload/drop in multiple documents, and delete documents. It appears as if I'm uploading one of each document.

After submitting the submission, those documents were duplicated and appears as multiple documents. Here is the test submission: https://test-apply.opentech.fund/apply/submissions/261/

**Comment by frjo** [comment]
If you upload the same file multiple times you get multiple versions of it. It does not get replaced with the new version as one might expect. @sks444 Any ideas about this?

**Comment by fourthletter** [comment]
I uploaded those documents exactly once.

**Comment by frjo** [comment]
@fourthletter Uploading a single version of a file and having it turn in to two or more I can't replicate. Can you please try it on a new submission and document the steps you take?

**Comment by fourthletter** [comment]
Duplication occurs when I '**dragged and dropped**' into the multiple upload section. This duplication did NOT occur when I used the file upload button.

**Comment by sks444** [comment]
> If you upload the same file multiple times you get multiple versions of it. It does not get replaced with the new version as one might expect. @sks444 Any ideas about this?

Feels like this validation should be done at upstream, I have opened an [issue](https://github.com/mbraak/django-file-form/issues/323) at django-file-form. I also tested other drag and drop uploaders, few of them have this validation and few of them doesn't. So it might be okay to go ahead with this and update django-file-form version when it is fixed at upstream, Or we could try to fix it there by ourselves before merging this pr.


>After submitting the submission, those documents were duplicated and appears as multiple documents. Here is the test submission: https://test-apply.opentech.fund/apply/submissions/261/

This was happening when submitting edit forms, but this should be fixed in last commits if that is deployed to test. I am not able to reproduce this locally. Some steps on how to reproduce this on a new submission would be helpful.

**Comment by frjo** [comment]
@fourthletter I can not replicate the duplication. Can you please test with a new submission and document the steps?

@sks444 Thanks for creating the issue for django-file-form on this. I think your approach sounds good.

**Comment by ei8fdb** [comment]
@frjo I'm not sure if this replicates what @fourthletter has seen. There are 2 issues here:
1. the applicant does not see in the widget UI all the files they are uploading
2. a bug where all acceptable files are not viewable in the Submissions (apply/submissions/) UI - only the last file.

Are they connected? I'm not sure.

### Explanation of issue
When an applicant uploads more than one acceptable file, they only ever see one file listed in the widget UI (the *last* file - odt-test-document.odt). This is confusing to them as they do not know if the other files were uploaded.

When a reviewer goes to the Submissions (apply/submissions/) UI, they only see the *last* file that the user uploaded. This must be a bug.

### Files uploaded
doc-test-document.doc
ods-test-document.ods
odt-test-document.odt

### What does the Hypha Applicant expect to see?
1. When I dragged and dropped the first file onto the upload widget, it would begin the file upload
2. While the uploading was happening I would see a progress bar
3. When the upload was finished I'd see the file shown as uploaded
4. When I dragged and dropped the second file onto the upload widget, it would begin the file upload
5. While the uploading was happening I would see a progress bar
6. When the upload was finished I'd see both filenames listed as uploaded

This would repeat for every file uploaded.

### What does the Hypha Applicant actually see?
1-5 above happened, but 6. did not. 

I only saw the *last* file listed. This means the user does not know what happened to the first file uploaded.

### What does the Hypha reviewer see?
In the backend for application name "Test 6. Upload multiple documents." (submission 276) the review only sees the *last* file uploaded.

### Example
[You can see if in this video (00:49 - 00:59)](https://files.apps.ei8fdb.org/Test%206.%20Upload%20multiple%20accepted%20files%20-%20UI%20only%20shows%20last%20file.webm)

**Comment by frjo** [comment]
@ei8fdb It actually works as expected. You have by mistake picked "File field" instead of "Multiple file field" for your application forms.

This is a very easy mistake to make and it has happened several times for OTF.

These two field really should be merged in to one site field type. On this it should be a setting where you specify how many files the user should be able to upload (1-10, unlimited or something like that.)

**Comment by frjo** [comment]
Added https://github.com/OpenTechFund/hypha/issues/2102
